### PR TITLE
blobstore: various fixes & cleanups

### PR DESCRIPTION
**Pull request by slimsag**
* Use `url.QueryEscape` for object names on disk, instead of a hash, so that it is reversible and we can use prefix matching on objects filenames when implementing the S3 List API.
* Simplify `multipartUploadMu` lock, clarify how mutexes & file closing behavior is designed.
* Fix `BLOBSTORE_DATA_DIR` initialization in single-binary mode (was read before single binary had a chance to set it)
* Improve filesystem atomics usage based on Keegan's feedback
* Add a `.tmp` suffix to tmp files so that we can clean them up later
* Refactor S3 HTTP routes into individual functions for cleaner codebase

Addresses much (but not all) of the feedback I got in https://github.com/sourcegraph/sourcegraph/pull/46972 - the following will be implemented in future PRs:

* S3 List API for listing objects (to make `ExpireObjects` test work)
* Background goroutine to cleanup tmp files and dead multipart uploads.

## Test plan

```sh
go test ./cmd/blobstore/...
```

This implementation of blobstore is not yet used anywhere outside of the Sourcegraph App, so there's little risk.

### CODE REVIEW

File: cmd/blobstore/internal/blobstore/blobstore.go [hunk 1]
```diff
@@ -355,6 +355,7 @@ func (s *Service) getObject(ctx context.Context, bucketName, objectName string)
 	defer bucketLock.RUnlock()
 
 	// Read the object
+	// Note that we return an io.ReadCloser here, so f.Close is intentionally NOT called.
```
Comment (efritz) [inline cmd/blobstore/internal/blobstore/blobstore.go:168]: That was my confusion on the previous PR. Feel free to leave it but I don't think this code was unclear in the first place (I was just tired).

File: cmd/blobstore/internal/blobstore/blobstore.go [hunk 2]
```diff
@@ -30,11 +30,11 @@ type Service struct {
...
 
-	initOnce          sync.Once
-	bucketLocksMu     sync.Mutex
-	bucketLocks       map[string]*sync.RWMutex
-	multipartUploadMu sync.Mutex
-	MockObjectAge     map[string]time.Time
+	initOnce              sync.Once
+	bucketLocksMu         sync.Mutex
+	bucketLocks           map[string]*sync.RWMutex
+	mutatePendingUploadMu sync.Mutex
```
Comment (efritz) [inline cmd/blobstore/internal/blobstore/blobstore.go:34]: q: Does this make it so that only one multipart upload can be serviced at a time?

**Review by efritz — APPROVED** [review]

File: cmd/blobstore/internal/blobstore/blobstore.go (hunk 2 above)
Comment (slimsag) [reply cmd/blobstore/internal/blobstore/blobstore.go:34]: No, this mutex only protects the *metadata* about a multipart upload. i.e. which part numbers it has, whether it is in progress, etc.

The actual process of uploading data once the metadata has been updated is fully parallel.

File: cmd/blobstore/internal/blobstore/s3_routes.go [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+			if r.URL.Query().Has("uploadId") {
+				return s.serveAbortMultipartUpload(w, r, bucketName, objectName)
+			}
+			return s.serveDeleteObject(w, r, bucketName, objectName)
+		}
+	}
+	return errors.Newf("unsupported method: %s request: %s", r.Method, r.URL)
+}
+
+// PUT /<bucket>
```
Comment (eseliger) [inline cmd/blobstore/internal/blobstore/s3_routes.go:58]: appreciate this code restructure!

**Review by eseliger — APPROVED** [review]

**Pull request merged**